satellite/gracefulexit/endpoint.go

```diff
@@ -344,60 +293,24 @@ func (endpoint *Endpoint) doProcess(stream processStream) (err error) {
 		if !morePiecesFlag && pendingCount == 0 {
 			processMu.Unlock()
 
-			exitStatusRequest := &overlay.ExitStatusRequest{
-				NodeID:         nodeID,
-				ExitFinishedAt: time.Now().UTC(),
-			}
-
 			progress, err := endpoint.db.GetProgress(ctx, nodeID)
 			if err != nil {
 				return rpcstatus.Error(rpcstatus.Internal, err.Error())
 			}
 
-			var transferMsg *pb.SatelliteMessage
 			mon.IntVal("graceful_exit_final_pieces_failed").Observe(progress.PiecesFailed)
 			mon.IntVal("graceful_exit_final_pieces_succeess").Observe(progress.PiecesTransferred)
 			mon.IntVal("graceful_exit_final_bytes_transferred").Observe(progress.BytesTransferred)
 
-			processed := progress.PiecesFailed + progress.PiecesTransferred
-			if processed > 0 {
-				mon.IntVal("graceful_exit_successful_pieces_transfer_ratio").Observe(progress.PiecesTransferred / processed)
-			}
-
-			// check node's exiting progress to see if it has failed passed max failure threshold
-			if processed > 0 && float64(progress.PiecesFailed)/float64(processed)*100 >= float64(endpoint.config.OverallMaxFailuresPercentage) {
-
-				exitStatusRequest.ExitSuccess = false
-				transferMsg, err = endpoint.getFinishedMessage(ctx, endpoint.signer, nodeID, exitStatusRequest.ExitFinishedAt, exitStatusRequest.ExitSuccess, pb.ExitFailed_OVERALL_FAILURE_PERCENTAGE_EXCEEDED)
-				if err != nil {
-					return rpcstatus.Error(rpcstatus.Internal, err.Error())
-				}
-			} else {
-				exitStatusRequest.ExitSuccess = true
-				transferMsg, err = endpoint.getFinishedMessage(ctx, endpoint.signer, nodeID, exitStatusRequest.ExitFinishedAt, exitStatusRequest.ExitSuccess, -1)
-				if err != nil {
-					return rpcstatus.Error(rpcstatus.Internal, err.Error())
-				}
-			}
-			if exitStatusRequest.ExitSuccess {
-				mon.Meter("graceful_exit_success").Mark(1)
-			} else {
-				mon.Meter("graceful_exit_fail_max_failures_percentage").Mark(1)
-			}
-			_, err = endpoint.overlaydb.UpdateExitStatus(ctx, exitStatusRequest)
+			// update exit status
+			exitStatusRequest, err := endpoint.generateExitStatusRequest(ctx, progress)
 			if err != nil {
 				return rpcstatus.Error(rpcstatus.Internal, err.Error())
 			}
 
-			err = stream.Send(transferMsg)
+			err = endpoint.handleFinished(ctx, stream, exitStatusRequest, pb.ExitFailed_OVERALL_FAILURE_PERCENTAGE_EXCEEDED)
```

Comment:
how do we know it definitely failed due to percentage exceeded here?

Reply:
I think it's because that's the only reason it would fail if it has finished transfer all pieces. I could be wrong tho

Reply:
`generateExitStatusRequest` should return a `pb.ExitFailed_Reason`, because this call to `handleFinished` could still be an exit success, in which case we want to pass in `-1` instead of `pb.ExitFailed_OVERALL_FAILURE_PERCENTAGE_EXCEEDED`